src/radiation/mocmc.cpp

```diff
@@ -0,0 +1,1100 @@
+  const int &nx_k = pmb->cellbounds.ncellsk(IndexDomain::interior);
+  const Real &dx_i = pmb->coords.dx1f(pmb->cellbounds.is(IndexDomain::interior));
+  const Real &dx_j = pmb->coords.dx2f(pmb->cellbounds.js(IndexDomain::interior));
+  const Real &dx_k = pmb->coords.dx3f(pmb->cellbounds.ks(IndexDomain::interior));
+  const Real &minx_i = pmb->coords.x1f(ib.s);
+  const Real &minx_j = pmb->coords.x2f(jb.s);
+  const Real &minx_k = pmb->coords.x3f(kb.s);
+
+  // Microphysics
+  auto opac = pmb->packages.Get("opacity");
+  const auto d_opac = opac->template Param<Opacity>("d.opacity");
+  StateDescriptor *eos = pmb->packages.Get("eos").get();
+
+  // auto &dnsamp = rc->Get(mocmc_internal::dnsamp);
+  std::vector<std::string> variables{pr::J,           pr::H,  pf::density, pf::velocity,
+                                     pf::temperature, pf::ye, im::dnsamp};
+  PackIndexMap imap;
+  auto v = rc->PackVariables(variables, imap);
+
+  auto pJ = imap.GetFlatIdx(pr::J);
+  auto pH = imap.GetFlatIdx(pr::H);
+  auto pdens = imap[pf::density].first;
+  auto pv = imap.GetFlatIdx(fluid_prim::velocity);
+  auto pT = imap[pf::temperature].first;
+  auto pye = imap[pf::ye].first;
+  auto dn = imap[im::dnsamp].first;
+
+  const int nblock = v.GetDim(5);
+  PARTHENON_REQUIRE_THROWS(nblock == 1, "Packing not currently supported for swarms");
+
+  const int nsamp_per_zone = rad->Param<int>("nsamp_per_zone");
+  int nsamp_tot = 0;
+
+  // Fill nsamp per zone per species and sum over zones
+  // TODO(BRR) make this a separate function and use it to update dnsamp which is then
+  // used to decide whether to refine/derefine
+  parthenon::par_reduce(
+      parthenon::loop_pattern_mdrange_tag, "MOCMC::Init::NumSamples", DevExecSpace(), 0,
+      nblock - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, int &nsamp) {
+        Real Jtot = 0.;
+        for (int s = 0; s < 3; s++) {
+          Jtot += v(b, pJ(s), k, j, i);
+        }
+        v(b, dn, k, j, i) =
+            get_nsamp_per_zone(k, j, i, geom, v(b, pdens, k, j, i), v(b, pT, k, j, i),
+                               v(b, pye, k, j, i), Jtot, nsamp_per_zone);
+        nsamp += v(b, dn, k, j, i);
+      },
+      Kokkos::Sum<int>(nsamp_tot));
+
+  ParArrayND<int> new_indices;
+  auto new_mask = swarm->AddEmptyParticles(nsamp_tot, new_indices);
+
+  // Calculate array of starting index for each zone to compute particles
+  ParArrayND<int> starting_index("Starting index", nx_k, nx_j, nx_i);
+  auto starting_index_h = starting_index.GetHostMirror();
+  auto dN = rc->Get(im::dnsamp).data;
+  auto dN_h = dN.GetHostMirrorAndCopy();
+  int index = 0;
+  for (int k = 0; k < nx_k; k++) {
+    for (int j = 0; j < nx_j; j++) {
+      for (int i = 0; i < nx_i; i++) {
+        starting_index(k, j, i) = index;
+        index += static_cast<int>(dN_h(k + kb.s, j + jb.s, i + ib.s));
+      }
+    }
+  }
+  starting_index.DeepCopy(starting_index_h);
```

Comment:
This looks like it could be very slow for large meshblocks. I wonder if there's a way to do this kind thing in parallel or on device? Probably not. But maybe worth engaging with the Kokkos people to see if they have a solution.

Reply:
Yeah I would like to find a better way to do this (it's an issue inside the parthenon particles as well)

Reply:
isn't this just a prefix sum which you can do with a `Kokkos::parallel_scan`?

Reply:
Yes. Now to add `par_scan` to parthenon (I left in an implementation that works on my CPU, but commented out for now)